### PR TITLE
Config parsing

### PR DESCRIPTION
**Pull request by enricoliscio**
Add option to parse an .ini config file. For now supports only anchor parameters.
Follows from suggestion of @yecharlie.

### CODE REVIEW

File: keras_retinanet/bin/train.py [hunk 1]
```diff
@@ -89,12 +90,14 @@ def create_models(backbone_retinanet, num_classes, weights, multi_gpu=0, freeze_
         weights            : The weights to load into the model.
         multi_gpu          : The number of GPUs to use for training.
         freeze_backbone    : If True, disables learning for the backbone.
+        config             : Config parameters, None indicates the default configuration.
```
Comment (vcarpani) [inline keras_retinanet/bin/train.py]: Alignment ;)

File: keras_retinanet/bin/train.py [hunk 2]
```diff
@@ -205,6 +212,7 @@ def create_generators(args, preprocess_image):
     """
     common_args = {
         'batch_size'       : args.batch_size,
+        'config'           : args.config,
```
Comment (vcarpani) [inline keras_retinanet/bin/train.py]: Alignment ;)

File: keras_retinanet/bin/train.py [hunk 3]
```diff
@@ -380,22 +388,23 @@ def csv_list(string):
...
+    parser.add_argument('--no-evaluation',    help='Disable per epoch evaluation.', dest='evaluation', action='store_false')
+    parser.add_argument('--freeze-backbone',  help='Freeze training of backbone layers.', action='store_true')
     parser.add_argument('--random-transform', help='Randomly transform image and annotations.', action='store_true')
-    parser.add_argument('--image-min-side', help='Rescale the image so the smallest side is min_side.', type=int, default=800)
-    parser.add_argument('--image-max-side', help='Rescale the image if the largest side is larger than max_side.', type=int, default=1333)
-    parser.add_argument('--weighted-average',   help='Compute the mAP using the weighted average of precisions among classes.', action='store_true')
+    parser.add_argument('--image-min-side',   help='Rescale the image so the smallest side is min_side.', type=int, default=800)
+    parser.add_argument('--image-max-side',   help='Rescale the image if the largest side is larger than max_side.', type=int, default=1333)
+    parser.add_argument('--config',           help='Path to a configuration parameters .ini file.', default=None)
```
Comment (vcarpani) [inline keras_retinanet/bin/train.py]: Alignment ;)

File: keras_retinanet/bin/train.py [hunk 4]
```diff
@@ -425,20 +438,24 @@ def main(args=None):
...
+            anchor_params = parse_anchor_parameters(args.config)
+            prediction_model = retinanet_bbox(model=model, anchor_params=anchor_params)
+        else:
+            prediction_model = retinanet_bbox(model=model)
     else:
         weights = args.weights
         # default to imagenet if nothing else is specified
         if weights is None and args.imagenet_weights:
             weights = backbone.download_imagenet()
-
```
Comment (vcarpani) [inline keras_retinanet/bin/train.py]: Why did you remove this line?

File: keras_retinanet/models/__init__.py [hunk 5]
```diff
@@ -70,6 +71,7 @@ def load_model(filepath, backbone_name='resnet50', convert=False, nms=True, clas
         convert               : Boolean, whether to convert the model to an inference model.
         nms                   : Boolean, whether to add NMS filtering to the converted model. Only valid if convert=True.
         class_specific_filter : Whether to use class specific filtering or filter for the best scoring class only.
```
Comment (vcarpani) [inline keras_retinanet/models/__init__.py]: Found one alignment  error from me, but it can be fixed in this PR

File: keras_retinanet/models/retinanet.py [hunk 6]
```diff
@@ -311,10 +282,10 @@ def retinanet(
 
 def retinanet_bbox(
     model                 = None,
-    anchor_parameters     = AnchorParameters.default,
     nms                   = True,
     class_specific_filter = True,
```
Comment (vcarpani) [inline keras_retinanet/models/retinanet.py]: Same as before, my error

File: keras_retinanet/models/retinanet.py [hunk 7]
```diff
@@ -324,10 +295,10 @@ def retinanet_bbox(
 
     Args
         model                 : RetinaNet model to append bbox layers to. If None, it will create a RetinaNet model using **kwargs.
-        anchor_parameters     : Struct containing configuration for anchor generation (sizes, strides, ratios, scales).
         nms                   : Whether to use non-maximum suppression for the filtering step.
         class_specific_filter : Whether to use class specific filtering or filter for the best scoring class only.
```
Comment (vcarpani) [inline keras_retinanet/models/retinanet.py]: Same as before, my error

File: keras_retinanet/preprocessing/generator.py [hunk 8]
```diff
@@ -77,6 +79,7 @@ def __init__(
         self.compute_anchor_targets = compute_anchor_targets
         self.compute_shapes         = compute_shapes
         self.preprocess_image       = preprocess_image
+        self.config                 = config
```
Comment (vcarpani) [inline keras_retinanet/preprocessing/generator.py]: Alignment ;)

**Comment by vcarpani** [comment]
@enricoliscio also check why the build is failing, probably coding style problems

File: keras_retinanet/bin/evaluate.py [hunk 9]
```diff
@@ -93,17 +97,18 @@ def parse_args(args):
...
+    parser.add_argument('--backbone',         help='The backbone of the model.', default='resnet50')
+    parser.add_argument('--gpu',              help='Id of the GPU to use (as reported by nvidia-smi).')
+    parser.add_argument('--score-threshold',  help='Threshold on score to filter detections with (defaults to 0.05).', default=0.05, type=float)
+    parser.add_argument('--iou-threshold',    help='IoU Threshold to count for a positive detection (defaults to 0.5).', default=0.5, type=float)
+    parser.add_argument('--max-detections',   help='Max Detections per image (defaults to 100).', default=100, type=int)
+    parser.add_argument('--save-path',        help='Path for saving images with detections (doesn\'t work for COCO).')
+    parser.add_argument('--image-min-side',   help='Rescale the image so the smallest side is min_side.', type=int, default=800)
+    parser.add_argument('--image-max-side',   help='Rescale the image if the largest side is larger than max_side.', type=int, default=1333)
+    parser.add_argument('--config',           help='Path to a configuration parameters .ini file.', default=None)
```
Comment (hgaiser) [inline keras_retinanet/bin/evaluate.py]: `None` is already the default, no need to define it.

File: keras_retinanet/bin/evaluate.py [hunk 10]
```diff
@@ -126,12 +131,20 @@ def main(args=None):
...
+    if args.config:
+        args.config = read_parameters_file(args.config)
+
     # create the generator
     generator = create_generator(args)
 
     # load the model
     print('Loading model, this may take a second...')
-    model = models.load_model(args.model, backbone_name=args.backbone, convert=args.convert_model)
+    if args.config and 'anchor_parameters' in args.config:
```
Comment (hgaiser) [inline keras_retinanet/bin/evaluate.py]:
```python
anchor_params = None
if args.config and 'anchor_parameters' in args.config:
    anchor_params = parse_anchor_parameters(args.config)
model = models.load_model(args.model, backbone_name=args.backbone, convert=args.convert_model, anchor_params=anchor_params)
```

Saves a bit of code duplication.

File: keras_retinanet/bin/train.py [hunk 11]
```diff
@@ -109,7 +112,11 @@ def create_models(backbone_retinanet, num_classes, weights, multi_gpu=0, freeze_
         training_model = model
 
     # make prediction model
-    prediction_model = retinanet_bbox(model=model)
+    if config and 'anchor_parameters' in config:
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: Same thing here.

File: keras_retinanet/bin/train.py [hunk 12]
```diff
@@ -425,7 +438,11 @@ def main(args=None):
         print('Loading model, this may take a second...')
         model            = models.load_model(args.snapshot, backbone_name=args.backbone)
         training_model   = model
-        prediction_model = retinanet_bbox(model=model)
+        if args.config and 'anchor_parameters' in args.config:
```
Comment (hgaiser) [inline keras_retinanet/bin/train.py]: And here.

File: keras_retinanet/preprocessing/generator.py [hunk 13]
```diff
@@ -238,7 +241,11 @@ def compute_inputs(self, image_group):
         return image_batch
 
     def generate_anchors(self, image_shape):
-        return anchors_for_shape(image_shape, shapes_callback=self.compute_shapes)
+        if self.config and 'anchor_parameters' in self.config:
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: And here.

File: keras_retinanet/utils/anchors.py [hunk 14]
```diff
@@ -212,8 +232,10 @@ def anchors_for_shape(
     # compute anchors over all pyramid levels
     all_anchors = np.zeros((0, 4))
     for idx, p in enumerate(pyramid_levels):
-        anchors         = generate_anchors(base_size=sizes[idx], ratios=ratios, scales=scales)
-        shifted_anchors = shift(image_shapes[idx], strides[idx], anchors)
+        anchors         = generate_anchors(base_size=anchor_params.sizes[idx],
```
Comment (hgaiser) [inline keras_retinanet/utils/anchors.py]: please align like:
```python
foo(
    bar=0,
    baz=1,
)
```

File: keras_retinanet/utils/config.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+limitations under the License.
+"""
+
+import configparser
+import numpy as np
+import keras
+from ..utils.anchors import AnchorParameters
+
+
+def read_parameters_file(parameters_path):
```
Comment (hgaiser) [inline keras_retinanet/utils/config.py]: Let's rename this to `read_config_file(config_path)`.

File: keras_retinanet/utils/config.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    return config
+
+
+def parse_anchor_parameters(config):
+    ratios  = []
+    scales  = []
+    sizes   = []
+    strides = []
+    for ratio in config['anchor_parameters']['ratios'].split(' '):
+        ratios.append(float(ratio))
```
Comment (hgaiser) [inline keras_retinanet/utils/config.py]: I think this can be simplified to:
```python
ratios = list(map(float, config['anchor_parameters']['ratios'].split(' ')))
```

**Comment by adreo00** [comment]
Wouldn't you also want to update debug.py to read from the .ini config file so that the proper anchor ratios will be displayed with the --anchors flag? I've made a pull request [here](https://github.com/yecharlie/keras-retinanet/pull/1) to add it. 

This is my first pull request so if this is not a good idea or it's in the wrong place, I'd love to hear feedback.

**Comment by enricoliscio** [comment]
Hey,
Thank you for your note, I had forgotten about debug.py. Unfortunately, that link leads me to another pull request with a different content. Please, branch out from the current branch (config-parsing, where the flag is --config and not --anchors, you probably branched out from an older version), and make a pull request to merge in config-parsing.

**Comment by adreo00** [comment]
@enricoliscio 
PR #688 should fix this

**Comment by hgaiser** [comment]
Looks good to me !
